### PR TITLE
Add one wallet endpoints

### PR DESCRIPTION
**Pull request by fuzzyami**

### CODE REVIEW

File: services/horizon/internal/init_web.go [hunk 1]
```diff
@@ -103,9 +103,14 @@ func initWebActions(app *App) {
 			r.Get("/offers", OffersByAccountAction{}.Handle)
 			r.Get("/trades", TradeIndexAction{}.Handle)
 			r.Get("/data/{key}", DataShowAction{}.Handle)
+			// Linked-Accounts actions:
+			r.Get("/aggregate_balance", AggregateBalanceAction{}.Handle)
+			r.Get("/controlled_balances", ControlledBalancesAction{}.Handle)
```
Comment (oryband) [inline services/horizon/internal/init_web.go]: 1. this isn't python world and no one is using underscore naming
1. if you can, use a single action word instead of two or more
1. search for RESTful API endpoint naming convention and pick something that makes sense

File: protocols/horizon/main.go [hunk 2]
```diff
@@ -451,3 +451,43 @@ func MustKeyTypeFromAddress(address string) string {
 
 	return ret
 }
+
+type AggregateBalanceRecord struct {
+	Id               string `json:"id"`
```
Comment (oryband) [inline protocols/horizon/main.go]: there's also `account_id` used in Horizon responses. Check the difference and keep it consistent with the rest of the responses Horizon does.

File: protocols/horizon/main.go [hunk 3]
```diff
@@ -451,3 +451,43 @@ func MustKeyTypeFromAddress(address string) string {
...
+		Precedes   hal.Link `json:"precedes"`
+		Succeeds   hal.Link `json:"succeeds"`
+	} `json:"_links"`
+	Embeded struct {
+		Records []AggregateBalanceRecord `json:"records"`
+	} `json:"_embedded"`
+}
+
+type ControlledBalanceRecord struct {
+	Id      string `json:"id"`
```
Comment (oryband) [inline protocols/horizon/main.go]: possible `account_id` needed?

File: services/horizon/internal/actions_aggregate_balance.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/kinecosystem/go/services/horizon/internal/resourceadapter"
+	"github.com/kinecosystem/go/support/render/hal"
+)
+
+// Interface verifications
+var _ actions.JSONer = (*AggregateBalanceAction)(nil)
+
+// AggregateBalanceAction renders a account summary found by its address.
+type AggregateBalanceAction struct {
+	Action
```
Comment (oryband) [inline services/horizon/internal/actions_aggregate_balance.go:16]: if `Action` is a struct, use a pointer. if it's an interface it's ok

File: services/horizon/internal/actions_aggregate_balance.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+func (action *AggregateBalanceAction) loadRecord() {
+	action.Err = action.CoreQ().AggregateBalanceByAccountId(&action.ControlledBalance)
+	if action.Err != nil {
+		return
+	}
+}
+
+func (action *AggregateBalanceAction) loadPage() {
+	var res core.AggregateBalance
+	resourceadapter.PopulateAggregateBalance(action.R.Context(), &res, action.ControlledBalance.AccountId, action.ControlledBalance.Balance)
```
Comment (oryband) [inline services/horizon/internal/actions_aggregate_balance.go]: break lines here like in the above functions

File: services/horizon/internal/actions_aggregate_balance_test.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+	// existing account
+	w := ht.Get(
+		"/accounts/GBRPYHIL2CI3FNQ4BXLFMNDLFJUNPU2HY3ZMFSHONUCEOASW7QC7OX2H/aggregate_balance",
+	)
+	if ht.Assert.Equal(200, w.Code) {
+		var aggregateBalances horizon.AggregatedBalances
+		err := json.Unmarshal(w.Body.Bytes(), &aggregateBalances)
+		fmt.Printf("%+v\n", aggregateBalances)
+		fmt.Printf("%+s\n", w.Body)
```
Comment (oryband) [inline services/horizon/internal/actions_aggregate_balance_test.go]: remove `import "fmt"` and all usages in tests. use only test log prints.

File: services/horizon/internal/actions_aggregate_balance_test.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+		ht.Require.NoError(err)
+		ht.Assert.Equal("GBRPYHIL2CI3FNQ4BXLFMNDLFJUNPU2HY3ZMFSHONUCEOASW7QC7OX2H", aggregateBalances.Embeded.Records[0].Id)
+		ht.Assert.Equal(999999996999999700, aggregateBalances.Embeded.Records[0].AggregateBalance)
+	}
+
+	// missing account
+	w = ht.Get("/accounts/GDBAPLDCAEJV6LSEDFEAUDAVFYSNFRUYZ4X75YYJJMMX5KFVUOHX46SQ/aggregate_balance")
+	ht.Assert.Equal(404, w.Code)
```
Comment (oryband) [inline services/horizon/internal/actions_aggregate_balance_test.go]: status codes are consted in golang stdlib if you'd like. you can use `net/http/status.StatusNotFound`: https://golang.org/src/net/http/status.go

File: services/horizon/internal/actions_controlled_balances_test.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
+package horizon
+
+import (
+	"encoding/json"
+	//"fmt"
```
Comment (oryband) [inline services/horizon/internal/actions_controlled_balances_test.go]: remove

Comment (oryband) [reply services/horizon/internal/actions_controlled_balances_test.go]: run `golint`, `govet`, `goerrcheck`, and also `gometalinter` which runs all of them

File: services/horizon/internal/actions_controlled_balances_test.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+	)
+	if ht.Assert.Equal(200, w.Code) {
+		var controlledBalances horizon.ControlledBalances
+		err := json.Unmarshal(w.Body.Bytes(), &controlledBalances)
+		ht.Require.NoError(err)
+		ht.Assert.Equal(0, len(controlledBalances.Embeded.Records))
+	}
+
+	// get controlled balance for andrew - should return 2 accounts (scott and andrew)
+	w1 := ht.Get(
```
Comment (oryband) [inline services/horizon/internal/actions_controlled_balances_test.go]: what does `w1` mean here? what does `w2` mean afterwards? use semantic naming

File: services/horizon/internal/db2/core/linked_accounts.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	sqltypes "database/sql"
+	"fmt"
+	sq "github.com/Masterminds/squirrel"
+	"github.com/kinecosystem/go/xdr"
+)
+
+// AggregateBaalanceByAccountId returns the aggregate balance by the master account id
+func (q *Q) AggregateBalanceByAccountId(dest *ControlledBalance) error {
+	sql := sq.Select("SUM(balance) aggbalance").From("accounts").Where(fmt.Sprintf("accountid = '%s' OR accountid IN (SELECT accountid FROM signers WHERE publickey = '%s')", dest.AccountId, dest.AccountId))
```
Comment (oryband) [inline services/horizon/internal/db2/core/linked_accounts.go]: you can break lines on periods in go

File: services/horizon/internal/db2/core/linked_accounts.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+		err = sqltypes.ErrNoRows // translates to 404 not found
+	}
+	return err
+}
+
+// ControlledBalancesByAccountId returns the list of controlled accounts with their respective balance
+func (q *Q) ControlledBalancesByAccountId(dest *[]*ControlledBalance, aid string) error {
+
+	// Ensure that the account exists
+	sql1 := sq.Select("count (*)").From("accounts a").Where(fmt.Sprintf("a.accountid='%s'", aid))
```
Comment (oryband) [inline services/horizon/internal/db2/core/linked_accounts.go]: `IsAccountExistsQuery := ...`

Comment (oryband) [reply services/horizon/internal/db2/core/linked_accounts.go]: see if we can avoid submitting a "count" query and just fetch all accounts at once or return an empty response with no error if there aren't any.

File: services/horizon/internal/db2/core/linked_accounts.go [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+	sql2 := sq.Select("a.accountid id, min(a.balance) balance").From("accounts a, signers s").
+		Where(fmt.Sprintf("s.publickey = '%s' AND (a.accountid = s.accountid OR a.accountid = s.publickey)", aid)).
+		GroupBy("a.accountid")
+	rows, err := q.Query(sql2)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	var balance_int int64
```
Comment (oryband) [inline services/horizon/internal/db2/core/linked_accounts.go]: this is not python. use go vet.

File: services/horizon/internal/db2/core/linked_accounts.go [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+		dest.Balance = xdr.Int64(result.NullableBalance.Int64)
+	} else {
+		err = sqltypes.ErrNoRows // translates to 404 not found
+	}
+	return err
+}
+
+// ControlledBalancesByAccountId returns the list of controlled accounts with their respective balance
+func (q *Q) ControlledBalancesByAccountId(dest *[]*ControlledBalance, aid string) error {
+
```
Comment (oryband) [inline services/horizon/internal/db2/core/linked_accounts.go]: extra line is extra

File: services/horizon/internal/db2/core/linked_accounts.go [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+	if result.NullableBalance.Valid {
+		dest.Balance = xdr.Int64(result.NullableBalance.Int64)
+	} else {
+		err = sqltypes.ErrNoRows // translates to 404 not found
+	}
+	return err
+}
+
+// ControlledBalancesByAccountId returns the list of controlled accounts with their respective balance
+func (q *Q) ControlledBalancesByAccountId(dest *[]*ControlledBalance, aid string) error {
```
Comment (oryband) [inline services/horizon/internal/db2/core/linked_accounts.go]: rename `dest` to `controlledBalances`

File: services/horizon/internal/db2/core/linked_accounts.go [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+		GroupBy("a.accountid")
+	rows, err := q.Query(sql2)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	var balance_int int64
+	for rows.Next() {
+		cb := ControlledBalance{}
```
Comment (oryband) [inline services/horizon/internal/db2/core/linked_accounts.go]: `var cb ControlledBalance`

File: services/horizon/internal/actions_controlled_balances.go [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+	return action.Err
+}
+
+func (action *ControlledBalancesAction) loadParams() {
+	action.Address = action.GetAddress("account_id", actions.RequiredParam)
+}
+
+func (action *ControlledBalancesAction) loadRecord() {
+	action.ControlledBalances = make([]*core.ControlledBalance, 0)
+	action.Err = action.CoreQ().ControlledBalancesByAccountId(&action.ControlledBalances, action.Address)
```
Comment (oryband) [inline services/horizon/internal/actions_controlled_balances.go]: always send slice by copy, not reference. same as interface. https://blog.golang.org/go-slices-usage-and-internals

File: services/horizon/internal/db2/core/linked_accounts.go [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+		err = rows.Scan(&cb.AccountId, &balance_int)
+		if err != nil {
+			return err
+		}
+		// convert the int64 to xdr
+		cb.Balance = xdr.Int64(balance_int)
+		// append the results into the slice:
+		*dest = append(*dest, &cb)
```
Comment (oryband) [inline services/horizon/internal/db2/core/linked_accounts.go]: slices are pointers already, no need to send pointer of pointer:
```golang
dest = append(dest, &cb)
```

File: services/horizon/internal/db2/core/main.go [hunk 18]
```diff
@@ -9,6 +9,24 @@ import (
 	"github.com/kinecosystem/go/xdr"
 )
 
+type AggregateBalance struct {
+	AccountId string    `json:"id"`
```
Comment (oryband) [inline services/horizon/internal/db2/core/main.go]: possibly missing `account_id`?

File: services/horizon/internal/db2/core/main.go [hunk 19]
```diff
@@ -9,6 +9,24 @@ import (
...
+	Balance   xdr.Int64 `json:"aggregate_balance"`
```
Comment (oryband) [inline services/horizon/internal/db2/core/main.go]: does it make sense to use `uint64` here or can it break the database and core?

File: services/horizon/internal/db2/core/main.go [hunk 20]
```diff
@@ -9,6 +9,24 @@ import (
...
+}
+
+func (r AggregateBalance) PagingToken() string {
+	return "0"
+}
```
Comment (oryband) [inline services/horizon/internal/db2/core/main.go:19]: document why you are returning a constant

File: services/horizon/internal/db2/core/main.go [hunk 21]
```diff
@@ -9,6 +9,24 @@ import (
 	"github.com/kinecosystem/go/xdr"
 )
 
+type AggregateBalance struct {
+	AccountId string    `json:"id"`
+	Balance   xdr.Int64 `json:"aggregate_balance"`
+}
+
+func (r AggregateBalance) PagingToken() string {
```
Comment (oryband) [inline services/horizon/internal/db2/core/main.go]: the receiver usually should be a pointer to the struct to avoid surprises.
```golang
func (r *AggregateBalance) ...
```
- read this: https://stackoverflow.com/questions/27775376/value-receiver-vs-pointer-receiver-in-golang and verify i'm not wrong
- implement this on all struct method implementations in this PR

File: services/horizon/internal/db2/core/main.go [hunk 22]
```diff
@@ -9,6 +9,24 @@ import (
...
+	return "0"
+}
+
+type ControlledBalance struct {
+	AccountId string    `json:"id"`
```
Comment (oryband) [inline services/horizon/internal/db2/core/main.go]: `account_id` ?

File: services/horizon/internal/init_web.go [hunk 23]
```diff
@@ -103,9 +103,14 @@ func initWebActions(app *App) {
 			r.Get("/offers", OffersByAccountAction{}.Handle)
 			r.Get("/trades", TradeIndexAction{}.Handle)
 			r.Get("/data/{key}", DataShowAction{}.Handle)
+			// Linked-Accounts actions:
```
Comment (oryband) [inline services/horizon/internal/init_web.go]: do a line space from the rest to avoid merge conflicts

Comment (oryband) [reply services/horizon/internal/init_web.go]: maybe also call this "one wallet" for readability

File: services/horizon/internal/resourceadapter/aggregate_balance.go [hunk 24]
```diff
@@ -0,0 +1,13 @@
+package resourceadapter
+
+import (
+	"context"
+	"github.com/kinecosystem/go/services/horizon/internal/db2/core"
+	"github.com/kinecosystem/go/xdr"
+)
+
+func PopulateAggregateBalance(ctx context.Context, dest *core.AggregateBalance, accountId string, balance xdr.Int64) {
```
Comment (oryband) [inline services/horizon/internal/resourceadapter/aggregate_balance.go]: when working with `xdr.` objects, do you need to work with copies or references?

File: services/horizon/internal/actions_aggregate_balance_test.go [hunk 25]
```diff
@@ -0,0 +1,31 @@
+package horizon
```
Comment (oryband) [inline services/horizon/internal/actions_aggregate_balance_test.go:1]: Test scenarios that we should test:
1. master account which doesn't exist - test both endpoints
1. master account which isn't linked to any other account - "
1. matser account which is linked to another account - "
1. matser account which is linked to two other accounts - "
1. matser account which is linked to three other accounts - "
1. test mass linking: you can use table test with the last three tests are, and test for linking 100+ accounts to a master account

edge cases:
1. account A linked to B, which also links back to A - test all endpoints
2. account A linked to B, account B linked to C - " + verify that A only shows B
3. cyclical: account A linked to B, B linked to C, C linked to A - " + verify A only shows B
4. repeat 1-3 but have A link to 3 normal accounts, and B as well to other 3 normal accounts. this is a "merge" of all test scenarios mention above this line.

File: services/horizon/internal/db2/core/linked_accounts.go [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+	"fmt"
+	sq "github.com/Masterminds/squirrel"
+	"github.com/kinecosystem/go/xdr"
+)
+
+// AggregateBaalanceByAccountId returns the aggregate balance by the master account id
+func (q *Q) AggregateBalanceByAccountId(dest *ControlledBalance) error {
+	sql := sq.Select("SUM(balance) aggbalance").From("accounts").Where(fmt.Sprintf("accountid = '%s' OR accountid IN (SELECT accountid FROM signers WHERE publickey = '%s')", dest.AccountId, dest.AccountId))
+	result := struct {
+		NullableBalance sqltypes.NullInt64 `db:"aggbalance"`
```
Comment (oryband) [inline services/horizon/internal/db2/core/linked_accounts.go:18]: should you use a pointer in the type here? verify with other implementations that already exist in the codebase

File: services/horizon/internal/db2/core/linked_accounts.go [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+
+	// Ensure that the account exists
+	sql1 := sq.Select("count (*)").From("accounts a").Where(fmt.Sprintf("a.accountid='%s'", aid))
+	var count int
+	err := q.Get(&count, sql1)
+	if err != nil {
+		return err
+	}
+	if count == 0 {
+		return sqltypes.ErrNoRows // translates to 404 not found
```
Comment (oryband) [inline services/horizon/internal/db2/core/linked_accounts.go]: standard in go is to put comments on top of code, not side by side

File: services/horizon/internal/db2/core/linked_accounts.go [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	// Get the data for the account
+	sql2 := sq.Select("a.accountid id, min(a.balance) balance").From("accounts a, signers s").
+		Where(fmt.Sprintf("s.publickey = '%s' AND (a.accountid = s.accountid OR a.accountid = s.publickey)", aid)).
+		GroupBy("a.accountid")
```
Comment (oryband) [inline services/horizon/internal/db2/core/linked_accounts.go:53]: i would let @doodyparizada look at the queries as well

**Review by oryband — CHANGES_REQUESTED** [review]

File: services/horizon/internal/init_web.go (hunk 1 above)
Comment (fuzzyami) [reply services/horizon/internal/init_web.go]: The current horizon version uses underscore in an endpoint. if they can, surely we can too. refer to https://github.com/stellar/go/blob/009e6e45c9f514e8e4f102461fbaefdf92b4f448/services/horizon/internal/web.go#L179

Comment (fuzzyami) [reply services/horizon/internal/init_web.go]: from what I gather, "id" is used to identify the page returned. in some cases, its the account id but in others, it might a tx hash or another identifier for the page. In other words, I'll need to use account_id instead of id.

File: protocols/horizon/main.go (hunk 2 above)
Comment (fuzzyami) [reply protocols/horizon/main.go]: from what I gather, "id" is used to identify the page returned. in some cases, its the account id but in others, it might a tx hash or another identifier for the page. In other words, I'll need to use account_id instead of id.

File: services/horizon/internal/actions_aggregate_balance.go (hunk 4 above)
Comment (fuzzyami) [reply services/horizon/internal/actions_aggregate_balance.go:16]: its done exactly this way in all the other actions.

File: services/horizon/internal/actions_controlled_balances_test.go (hunk 9 above)
Comment (fuzzyami) [reply services/horizon/internal/actions_controlled_balances_test.go]: this is Stellar's convention. I replaced with "resp" (although technically its a "request helper", whatever that is)

File: services/horizon/internal/db2/core/linked_accounts.go (hunk 12 above)
Comment (fuzzyami) [reply services/horizon/internal/db2/core/linked_accounts.go]: go vet had no issues with this. I renamed anyways.

File: services/horizon/internal/db2/core/main.go (hunk 19 above)
Comment (fuzzyami) [reply services/horizon/internal/db2/core/main.go]: Stellar uses int64 for balance.

File: services/horizon/internal/resourceadapter/aggregate_balance.go (hunk 24 above)
Comment (fuzzyami) [reply services/horizon/internal/resourceadapter/aggregate_balance.go]: AFAIK, copies.
